File: src/Butil/Bit.Butil/Scripts/bit-butil.ts
var butil = (function () {
    const _handlers = {};

    return {
        addEventListener,
        removeEventListener
    };

    function addEventListener(elementName, eventName, dotnetMethodName, dotnetListenerId, selectedMembers, options) {
        const handler = e => {
            DotNet.invokeMethodAsync('Bit.Butil', dotnetMethodName, dotnetListenerId, selectedMembers && selectedMembers.reduce((pre, cur) => (pre[cur] = e[cur], pre), {}));
        };

        _handlers[dotnetListenerId] = handler;

        window[elementName].addEventListener(eventName, handler, options);
    }

    function removeEventListener(elementName, eventName, dotnetListenerIds, options) {
        dotnetListenerIds.forEach(id => {
            const handler = _handlers[id];
            window[elementName].removeEventListener(eventName, handler, options);
        });
    }
}());

File: src/Butil/Bit.Butil/DOM/BitButil.cs
using System;
using Microsoft.JSInterop;

namespace Bit.Butil;

public static class BitButil
{
    private static bool _isInitialized;
    private static IJSRuntime _js = default!;

    public static void Init(IJSRuntime jsRuntime)
    {
        if (_isInitialized) return;

        _isInitialized = true;
        _js = jsRuntime;
    }

    internal static void AddEventListener(string elementName, string eventName, string dotnetMethodName, Guid dotnetListenerId, string[] selectedMembers, object? options = null)
    {
        var _ = _js.InvokeVoidAsync("butil.addEventListener", elementName, eventName, dotnetMethodName, dotnetListenerId, selectedMembers, options);
    }

    internal static void RemoveEventListener(string elementName, string eventName, Guid[] dotnetListenerIds, object? options = null)
    {
        var _ = _js.InvokeVoidAsync("butil.removeEventListener", elementName, eventName, dotnetListenerIds, options);
    }
}
